### ISSUE TITLE
data.forceArray applied on undefined

### ISSUE THREAD

**Issue by gbbirkisson** [enonic/lib-util#28]
In my opinion data.forceArray should return an empty array when passed undefined. i.e:

`forceArray(undefined) -> []`

not

`forceArray(undefined) -> [undefined]`

**Comment by hjelmevold** [comment]
I agree that this would make things easier, and the same goes for `null`

But I'm not sure if it's best to return `[]` or if it's better to return the original `undefined` (or in case of null: `null`)

Is there a good argument for why undefined or null should return an empty array? I have no preference.

Any suggestions, @ComLock ? @poi33 ?

**Comment by poi33** [comment]
I don't see the use for undefined -> [undefined] and null -> [null] 
So i would defently want them to return null or undefined in those cases. 
I could see them returning an emtpy array [], but this could mask/hide errors in the code, so i prefer them retruning the value instead.

**Comment by ComLock** [comment]
I agree with @gbbirkisson about undefined

```
forceArray(undefined) -> []
`
``

Null however is an actual value so
```
forceArray(null) -> [null]
```

Implememting this with array.filter(x => x) would be a bad idea because that would alter an array in which the number of elements is important.

```
[1, 0, -1, 'str', '', null, undefined, NaN, Infinity, () => {}]
`
``

**Comment by poi33** [comment]
Okay can someone add the force array with undefined returning an emtpy array?

**Comment by gbbirkisson** [comment]
I think ForceArray should always return an array. You should be able to do:

```javascript
if (content.data) {

  var maybeNull = content.data.someInput;
  forceArray(maybeNull).forEach(function (item, index) {
     // I wont get item == null
  });

}
```
ForceArray would then cover all input type occurrences, i.e:
* no values (null)
* 1 value
* n values

**Comment by ComLock** [comment]
In JS undefined means a variable has been declared but has not yet been assigned a value.
null is an assignment value. It can be assigned to a variable as a representation of no value.

Related thoughts:

forceArray is mostly used for working with node and content data.

Enonic XP does not store undefined or null.
Enonic XP stores an array with a single item as just the item.
I don't know what happens to an array with just a single null value in it.

But either way null is an actual value in js.

**Comment by ComLock** [comment]
```js
let variable = true;

// I believe this
variable = undefined;

// Is the same as
delete variable;

// While
let anotherVariable = true;

anotherVariable = null;

// Is NOT the same as
delete anotherVariable;
```

**Comment by rymsha** [comment]
Seems like forceArray(v) behaves very much like [].concat(v) .

**Comment by rymsha** [comment]
One doesn't need forceArray to get what @gbbirkisson specified either.
`[].concat(undefined || [])`  -> `[]`
`[].concat('a' || [])`  -> `['a']`
`[].concat(['a'] || [])`  -> `['a']`

**Comment by GlennRicaud** [comment]
Agreeing on the: forceArray(undefined) -> []
The null is more arguable. I would personally also expect as @ComLock  forceArray(null) -> [null]
